### PR TITLE
[feature] tasks

### PR DESCRIPTION
**Pull request by henriblancke**
Attempt to add [task](https://docs.snowflake.com/en/user-guide/tasks-intro.html) support.

Couple of notes:
- There is no support for the copy grants option on tasks as we're never cloning or replacing an existing task
- `after` only supports tasks that are in the same schema (as per the [docs](https://docs.snowflake.com/en/sql-reference/sql/create-task.html#optional-parameters))
- This PR doesn't implement task grants but we could make it part of it?
- The task is automatically started after creation

[Travis build](https://travis-ci.com/github/henriblancke/terraform-provider-snowflake/builds/168942892)

### CODE REVIEW

File: pkg/resources/task_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		"sql_statement": "select hi from hello",
+		"comment":       "wow comment",
+	}
+
+	d := schema.TestResourceDataRaw(t, resources.Task().Schema, in)
+	r.NotNil(d)
+
+	WithMockDb(t, func(db *sql.DB, mock sqlmock.Sqlmock) {
+		mock.ExpectExec(
+			`^CREATE TASK "test_db"."test_schema"."test_task" WAREHOUSE = much_warehouse COMMENT = 'wow comment' AS select hi from hello$`,
```
Comment (ryanking) [inline pkg/resources/task_test.go:38]: should the warehouse name be quoted?

File: pkg/snowflake/task.go [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+// ChangeWarehouse returns the sql that will change the warehouse for the task.
+func (tb *TaskBuilder) ChangeWarehouse(newWh string) string {
+	return fmt.Sprintf(`ALTER TASK %v SET WAREHOUSE = %v`, tb.QualifiedName(), newWh)
+}
+
+// ChangeSchedule returns the sql that will change the schedule for the task.
+func (tb *TaskBuilder) ChangeSchedule(newSchedule string) string {
+	return fmt.Sprintf(`ALTER TASK %v SET SCHEDULE = '%v'`, tb.QualifiedName(), newSchedule)
```
Comment (ryanking) [inline pkg/snowflake/task.go:161]: This (and the rest of the queries here) should probably use escaping - https://github.com/chanzuckerberg/terraform-provider-snowflake/blob/37ea8346c3ac656088a1cc47efc4f602c9966965/pkg/snowflake/escaping.go#L7

File: pkg/resources/task.go [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+	"schedule": &schema.Schema{
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "The schedule for periodically running the task. This can be a cron or interval in minutes.",
+	},
+	"session_parameters": &schema.Schema{
+		Type:        schema.TypeSet,
+		Elem:        &schema.Schema{Type: schema.TypeString},
+		Optional:    true,
+		Description: "Specifies a comma-separated list of session parameters to set for the session when the task runs. A task supports all session parameters.",
```
Comment (ryanking) [inline pkg/resources/task.go:53]: It is not clear to me how this is supposed to be used?  A set of strings, each of which has a comma-separated list of paremeters?

Would `schema.TypeMap` be clearer?

File: pkg/resources/task.go [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+
+	if data.HasChange("after") {
+		var (
+			q   string
+			err error
+		)
+		old, new := data.GetChange("after")
+
+		if old != "" {
+			q = builder.RemoveDependency(old.(string))
```
Comment (ryanking) [inline pkg/resources/task.go:354]: It seems from the [docs](https://docs.snowflake.com/en/sql-reference/sql/alter-task.html) that removing will suspend the task and "must be restarted manually". Does adding a new `after` restart the task?

File: pkg/resources/task.go (hunk 3 above)
Comment (henriblancke) [reply pkg/resources/task.go:53]: Yes that would make more sense! It would simplify the logic of removing session parameters as well! Thanks, didn't think of `schema.TypeMap`

File: pkg/resources/task.go (hunk 4 above)
Comment (henriblancke) [reply pkg/resources/task.go:354]: Hmm, good catch! I'll take a look

Comment (henriblancke) [reply pkg/resources/task.go:370]: I'm resuming the task every time `after` changes now.

**Comment by ryanking** [comment]
@henriblancke it appears there is another PR covering this -> https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/198

Can you all work together on this?

**Comment by chriskuchin** [comment]
@henriblancke @ryanking Happy to collaborate here the core of our 2 MRs seems very similar minus a few cosmetic differences. The biggest difference seems to be in mine I handle temporarily suspending the root task of a tree so that the child tasks can be altered and then resume it after the fact.

From the (docs)[https://docs.snowflake.com/en/sql-reference/sql/alter-task.html]:
```Modifying a standalone task requires that the task is suspended.

Modifying any task in a tree of tasks requires that the root task is suspended.
```

**Comment by henriblancke** [comment]
@chriskuchin yeah your handling of task dependencies seems much better. Having the option to enable/disable the task is also nice! One thing I noticed tho is that you're not exposing `session_parameters` in your PR? Maybe we should just add that and we're good to go? It's up to @ryanking on how you all want to handle this :)

**Comment by chriskuchin** [comment]
Yeah I got distracted and forgot about them. Happy to put a little more time into the PR if that's what we wanna do and add that

**Comment by ryanking** [comment]
> One thing I noticed tho is that you're not exposing `session_parameters` in your PR?

Perhaps we can move forward by merging @chriskuchin's PR then adding session_params as a separate PR?

**Comment by chriskuchin** [comment]
I nearly have the session_parameters piece done on my PR just testing and debugging now @ryanking

**Comment by henriblancke** [comment]
@ryanking @chriskuchin I actually went ahead and added the `enabled` flag and logic to suspend root or standalone tasks before modifying / creating a child task. I think my use of `defer` to reactivate the suspended root task might be a little cleaner and safer when anything else fails during modifying or creating a new task. That way we'll always re-enable the root task? Let me know what you all think...

**Comment by chriskuchin** [comment]
@henriblancke sounds good. Have you tested you session parameters? I had to modify them pretty heavily for them to actually work. Plus you weren't reading their state at all.

**Comment by henriblancke** [comment]
@chriskuchin it did test em, but maybe not enough 😄? Can you elaborate? What do you mean by not reading state? Thanks!

**Comment by chriskuchin** [comment]
You never read at least that I could find the parameters after you set them. The other problem I had was you were escaping the parameter values which caused them to be invalid as well as not wrapping them in quotes

File: pkg/resources/task.go [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+	err = data.Set("sql_statement", t.Definition)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateTask implements schema.CreateFunc
+func CreateTask(data *schema.ResourceData, meta interface{}) error {
```
Comment (chriskuchin) [inline pkg/resources/task.go:318]: @henriblancke you need root suspend logic in the create

File: pkg/resources/task.go [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+		dbSchema = taskID.SchemaName
+		name     = taskID.TaskName
+		builder  = snowflake.Task(name, database, dbSchema)
+	)
+
+	root, err := getRootTaskAndSuspend(data, meta)
+	if err != nil {
+		return err
+	}
+	defer resumeTask(root, meta)
```
Comment (chriskuchin) [inline pkg/resources/task.go:415]: @henriblancke you need to handle the case where the root was suspended

File: pkg/resources/task.go [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+			return errors.Wrapf(err, "error updating task state %v", data.Id())
+		}
+		data.SetPartial("enabled")
+	}
+
+	return ReadTask(data, meta)
+}
+
+// DeleteTask implements schema.DeleteFunc
+func DeleteTask(data *schema.ResourceData, meta interface{}) error {
```
Comment (chriskuchin) [inline pkg/resources/task.go:587]: @henriblancke you need to handle root task suspend/resume here as well

**Comment by henriblancke** [comment]
> You never read at least that I could find the parameters after you set them. The other problem I had was you were escaping the parameter values which caused them to be invalid as well as not wrapping them in quotes

yep I did just run into the quoting issue. will look into that

**Comment by chriskuchin** [comment]
Might be worthwhile to pull my acceptance test over into yours as it does a pretty good job of iterating a whole bunch of cases

File: pkg/resources/task.go (hunk 7 above)
Comment (henriblancke) [reply pkg/resources/task.go:587]: @chriskuchin it should be handled?

Comment (chriskuchin) [reply pkg/resources/task.go:587]: Yeah you need to suspend the root to delete a child task

File: pkg/resources/task.go (hunk 6 above)
Comment (henriblancke) [reply pkg/resources/task.go:415]: @chriskuchin check `getRootTask` and `getRootTaskAndSuspend`. I added some logic there to not return the root task builder when the root task is already suspended that way the logic here and in other places stays concise and simple? Let me know if you think it is misleading or makes the logic hard to follow.

**Comment by henriblancke** [comment]
> Might be worthwhile to pull my acceptance test over into yours as it does a pretty good job of iterating a whole bunch of cases

Agreed 👍

File: pkg/resources/task.go (hunk 6 above)
Comment (chriskuchin) [reply pkg/resources/task.go:415]: ahh nice I think though you are missing an edge case where you change multiple things including the state. i.e. i think if you were to alter some field plus change enabled -> false on a root node you would end up with enabled true at the end maybe?

File: pkg/resources/task.go (hunk 7 above)
Comment (henriblancke) [reply pkg/resources/task.go:587]: @chriskuchin thats what should be happening at line 600 here

File: pkg/resources/task.go [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+	)
+	if err != nil {
+		return err
+	}
+
+	database := taskID.DatabaseName
+	schema := taskID.SchemaName
+	name := taskID.TaskName
+
+	root, err := getRootTaskAndSuspend(data, meta)
```
Comment (henriblancke) [inline pkg/resources/task.go:600]: @chriskuchin here I'm suspending the root task before dropping the actual task on line 610

Comment (chriskuchin) [reply pkg/resources/task.go:600]: 🤦 I must be blind

File: pkg/resources/task.go (hunk 6 above)
Comment (henriblancke) [reply pkg/resources/task.go:415]: @chriskuchin how would you handle that? It depends in what order terraform executes changes and how it builds its dag? If there is some order through tf variables, it would most likely alter the root task first before going onwards to the children or am I thinking this all the wrong way?

File: pkg/resources/task.go (hunk 7 above)
Comment (chriskuchin) [reply pkg/resources/task.go:587]: yeah you are good I just missed those lines

File: pkg/resources/task.go (hunk 6 above)
Comment (chriskuchin) [reply pkg/resources/task.go:415]: you get a single nodes updates all at once. So if the node you are modifying is the root and currently enabled You will suspend it then if there is a change to suspend the node you are on further down the method call you will go suspend it. Then your defer will trigger and reenable it

Comment (chriskuchin) [reply pkg/resources/task.go:415]: The whole state machine around this took me forever and I really didn't love my code to handle it but my acceptance tests should exercise this case i believe

Comment (chriskuchin) [reply pkg/resources/task.go:415]: I think you may be able to solve the problem by if you handle a change in state you set the root to nil when you handle the node state change so your resume is a no op

Comment (henriblancke) [reply pkg/resources/task.go:415]: @chriskuchin added your acceptance tests

File: pkg/snowflake/task.go [hunk 9]
```diff
@@ -128,7 +128,7 @@ func (tb *TaskBuilder) Create() string {
 		sort.Strings(sortedKeys)
 
 		for _, k := range sortedKeys {
-			sp = append(sp, EscapeString(fmt.Sprintf(`%v = %v`, k, tb.session_parameters[k])))
+			sp = append(sp, fmt.Sprintf(`%v = %v`, k, tb.session_parameters[k]))
```
Comment (chriskuchin) [inline pkg/snowflake/task.go:131]: I think you also need to wrap the value in quotes since some of those values aren't valid ids

Comment (henriblancke) [reply pkg/snowflake/task.go:131]: It's difficult here because in the case of `CLIENT_TIMESTAMP_TYPE_MAPPING = TIMESTAMP_LTZ` for example, snowflake won't recognize that if you quote it? I figured I'd leave it up to the end user here to make sure they quote what needs to be quotes?

Comment (chriskuchin) [reply pkg/snowflake/task.go:131]: It will as long as they all caps it

Comment (chriskuchin) [reply pkg/snowflake/task.go:131]: Since most of the other ids in teh system are auto quoted in the provider I would prefer just quoting everything

File: pkg/resources/task.go (hunk 6 above)
Comment (henriblancke) [reply pkg/resources/task.go:415]: @chriskuchin I want to understand better how you handled this edge case and how you would prevent things from being executed in parallel? I think by default terraform executes 10 operations in parallel as it walks the graph. If there is no relation that helps tf define the order when it builds the graph, there is no way to prevent it from running both operations in your edge case at the same time? Let me know if I'm looking at this in the wrong way

Comment (henriblancke) [reply pkg/resources/task.go:415]: could people use `depends_on` to avoid this edge case?

Comment (chriskuchin) [reply pkg/resources/task.go:415]: There are 2 edge cases 1 that i don't think you can handle which is you concurrently modify 2 nodes, but can be prevented by using references from tasks for the after clause. and the second that is if you disable an enabled task on a node that is the root node then you will reactivate it with your defer

Comment (chriskuchin) [reply pkg/resources/task.go:415]: the operations from my testing happen on a node basis not a change in node basis. So you get all of a nodes changes in one call to the update method

File: pkg/resources/task.go [hunk 10]
```diff
@@ -0,0 +1,645 @@
...
+
+		Schema: taskSchema,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+// ReadTask implements schema.ReadFunc
+func ReadTask(data *schema.ResourceData, meta interface{}) error {
```
Comment (chriskuchin) [inline pkg/resources/task.go:244]: >>>What do you mean by not reading state? Thanks!

Here is ReadTask you need to read the parameters to be able to add them to the state

File: pkg/resources/task.go (hunk 6 above)
Comment (chriskuchin) [reply pkg/resources/task.go:415]: depends_on or using variable references would fix the first concurrent execution problem but doesn't fix the other. Have you run my acceptance tests against your code?

Comment (chriskuchin) [reply pkg/resources/task.go:415]: Also if it is easier I am happy to hop on a zoom for a quick conversation

File: pkg/resources/task.go (hunk 10 above)
Comment (henriblancke) [reply pkg/resources/task.go:244]: @chriskuchin does the `SHOW TASKS` return the `session_parameters`?

Comment (chriskuchin) [reply pkg/resources/task.go:244]: no you have to do another query you can see in my PR. It's a show parameters query

File: pkg/resources/task.go (hunk 6 above)
Comment (henriblancke) [reply pkg/resources/task.go:415]: @chriskuchin lets hop on [a zoom call](https://us04web.zoom.us/j/75028769614?pwd=b2JPZWF0UXBUN1pGTE5Gd1Q4UmtmUT09)

**Comment by henriblancke** [comment]
@chriskuchin any idea why [the acceptance test here](https://travis-ci.com/github/henriblancke/terraform-provider-snowflake/jobs/338886141) would be failing?

**Comment by chriskuchin** [comment]
They aren't even running they error out about not being configured in Travis. Are they passing for you locally?

**Comment by chriskuchin** [comment]
Ahh you have the acceptance tests hooked up to your fork... I have no idea that was never an error I saw when writing them or getting things working. Sounds like it's an error in the first step. You can enable tf_logs=debug to get more info but it's really noisy so might be hard. Can you run it locally so you run just that test with debug?

**Comment by chriskuchin** [comment]
So it seems to have something to do with your state. The error occurs during a state refresh so it's probably in the read function I would guess

**Comment by henriblancke** [comment]
@chriskuchin yeah just figured it out adding the id we get from snowflake during reading the task caused the error. It breaks `taskIDFromString`. I'm running into more problems tho now that this one is resolved

**Comment by chriskuchin** [comment]
ahh that makes sense I didn't follow that compound id pattern in My MR 👍 let me know if i can help

**Comment by ryanking** [comment]
> @ryanking not sure why the Acceptance tests aren't running in your CI but as long as those are passing the PR looks good to me

ok i will take a look

**Comment by henriblancke** [comment]
@ryanking they should be all green [here](https://travis-ci.com/github/henriblancke/terraform-provider-snowflake/builds/167959488)

File: pkg/resources/task.go [hunk 11]
```diff
@@ -0,0 +1,680 @@
...
+		dbSchema = taskID.SchemaName
+		name     = taskID.TaskName
+		builder  = snowflake.Task(name, database, dbSchema)
+	)
+
+	root, err := getRootTaskAndSuspend(data, meta)
+	if err != nil {
+		return err
+	}
+	defer resumeTask(root, meta)
```
[the remark this answers is not in the document]
Comment (henriblancke) [reply pkg/resources/task.go:446]: Yeah we'd want to root task to be re-enabled on a TF or config failure. We don't want it to be suspended indefinitely because of an error.

File: pkg/resources/task.go [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+		row := snowflake.QueryRow(db, q)
+		task, err := snowflake.ScanTask(row)
+
+		if err != nil && name != data.Get("name").(string) {
+			return nil, errors.Wrapf(err, "failed to locate the root node of: %v", name)
+		}
+
+		if task.Predecessors == nil {
+			log.Println(fmt.Sprintf("[DEBUG] found root task: %v", name))
+			// we only want to deal with suspending the root task when its enabled (started)
```
[the remark this answers is not in the document]
Comment (henriblancke) [reply pkg/resources/task.go:159]: Yeah we can rename it. What do you propsose? My aim was to avoid checking the same condition at the caller levels in multiple places.

[the remark this answers is not in the document]
Comment (ryanking) [reply pkg/resources/task.go:159]: Maybe getRootActiveTask ?

File: pkg/snowflake/task.go [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+	return strings.ToLower(t.State) == "started"
+}
+
+func (t *task) GetPredecessorName() string {
+	if t.Predecessors == nil {
+		return ""
+	}
+
+	pre := strings.Split(*t.Predecessors, ".")
+	name, _ := strconv.Unquote(pre[len(pre)-1])
```
Comment (chriskuchin) [inline pkg/snowflake/task.go:311]: Snowflake doesn't always wrap the name of the task in quotes and if there aren't quotes around the string then this errors and the result returned is an empty string

Comment (chriskuchin) [reply pkg/snowflake/task.go:311]:
```suggestion
		name, err := strconv.Unquote(pre[len(pre)-1])
	if err != nil {
		return pre[len(pre)-1]
	}

```

Comment (chriskuchin) [reply pkg/snowflake/task.go:311]: If you modify the root task's name to be all caps it will reproduce in the acceptance tests
